### ISSUE TITLE
Allow option to turn off validation for contained resources

### ISSUE THREAD

**Issue by jaison-b** [hapifhir/org.hl7.fhir.core#654]
Provide option to selectively turn off validation for contained resources.

Sample JSON:

```json
{
  "resourceType": "Bundle",
  "type": "transaction",
  "entry": [
    {
      "fullUrl": "Procedure/d3fe9844-9980-4273-81ae-b1ac6d45b4bf",
      "resource": {
        "resourceType": "Procedure",
        "id": "d3fe9844-9980-4273-81ae-b1ac6d45b4bf",
        "meta": {
          "profile": [
            "https://example.com/StructureDefinition/Procedure-v1"
          ]
        },
        "contained": [
          {
            "resourceType": "DiagnosticReport",
            "id": "1",
            "meta": {
              "profile": [
                "https://example.com/StructureDefinition/DiagnosticReport-v2"
              ]
            },
            "identifier": [
              {
                "system": "https://example.com/diagnosticReport",
                "value": "DiagnosticReport:3189671e-551d-4025-9077-3d26189b56ca:f24d3629-573b-470e-9110-61df35ecbdee:2079062e-0eb2-479c-942f-a7d5ff84bf0c:6f716159-c9fe-4825-8cab-8e94437cbcba:20210812004700"
              }
            ],
            "status": "final",
            "category": [
              {
                "coding": [
                  {
                    "system": "https://example.com/treatment",
                    "code": "Summary",
                    "display": "Summary"
                  }
                ]
              }
            ],
            "code": {
              "coding": [
                {
                  "system": "http://loinc.org",
                  "code": "LA17815-4",
                  "display": "Continuous cycling peritoneal dialysis"
                }
              ]
            },
            "subject": {
              "reference": "Patient/3189671e-551d-4025-9077-3d26189b56ca"
            },
            "effectiveDateTime": "2021-08-12T00:47:00+00:00",
            "result": [
              {
                "reference": "#2"
              }
            ]
          },
          {
            "resourceType": "Observation",
            "id": "2",
            "meta": {
              "profile": [
                "https://example.com/StructureDefinition/Observation-v1"
              ]
            },
            "identifier": [
              {
                "system": "https://example.com/observation",
                "value": "Observation:3189671e-551d-4025-9077-3d26189b56ca:f24d3629-573b-470e-9110-61df35ecbdee:2079062e-0eb2-479c-942f-a7d5ff84bf0c:6f716159-c9fe-4825-8cab-8e94437cbcba:20210812004700:BP"
              }
            ],
            "status": "final",
            "category": [
              {
                "coding": [
                  {
                    "system": "http://terminology.hl7.org/CodeSystem/observation-category",
                    "code": "vital-signs",
                    "display": "Vital Signs"
                  }
                ],
                "text": "Vital Signs"
              }
            ],
            "code": {
              "coding": [
                {
                  "system": "http://loinc.org",
                  "code": "8462-4",
                  "display": "Diastolic blood pressure"
                }
              ],
              "text": "Diastolic Blood Pressure"
            },
            "subject": {
              "reference": "Patient/3189671e-551d-4025-9077-3d26189b56ca"
            },
            "effectiveDateTime": "2021-08-12T00:47:00+00:00",
            "device": {
              "reference": "Device/a02efd01-6975-40e3-80af-d3eddbfe4560"
            },
            "component": [
              {
                "code": {
                  "coding": [
                    {
                      "system": "http://loinc.org",
                      "code": "8480-6",
                      "display": "Systolic blood pressure"
                    },
                    {
                      "system": "http://snomed.info/sct",
                      "code": "271649006",
                      "display": "Systolic blood pressure"
                    },
                    {
                      "system": "http://acme.org/devices/clinical-codes",
                      "code": "bp-s",
                      "display": "Systolic blood pressure"
                    }
                  ],
                  "text": "Systolic Blood Pressure"
                },
                "valueQuantity": {
                  "value": 120,
                  "unit": "mmHg",
                  "system": "http://unitsofmeasure.org"
                }
              },
              {
                "code": {
                  "coding": [
                    {
                      "system": "http://loinc.org",
                      "code": "8462-4",
                      "display": "Diastolic blood pressure"
                    }
                  ],
                  "text": "Diastolic Blood Pressure"
                },
                "valueQuantity": {
                  "value": 80,
                  "unit": "mmHg",
                  "system": "http://unitsofmeasure.org"
                }
              }
            ]
          }
        ],
        "identifier": [
          {
            "system": "https://example.com/procedure",
            "value": "Procedure:3189671e-551d-4025-9077-3d26189b56ca:f24d3629-573b-470e-9110-61df35ecbdee:2079062e-0eb2-479c-942f-a7d5ff84bf0c:6f716159-c9fe-4825-8cab-8e94437cbcba:20210812000000"
          }
        ],
        "status": "unknown",
        "category": {
          "coding": [
            {
              "system": "http://snomed.info/sct",
              "code": "718308002",
              "display": "Peritoneal dialysis care"
            }
          ]
        },
        "subject": {
          "reference": "Patient/3189671e-551d-4025-9077-3d26189b56ca"
        },
        "performedPeriod": {
          "start": "2021-08-12T00:47:00+00:00",
          "end": "2021-08-12T10:35:00+00:00"
        }
      },
      "request": {
        "method": "POST",
        "url": "Procedure"
      }
    }
  ]
}
```

In the sample above, we would like to only validate the parent procedure and not the contained resources.

@markiantorno

**Comment by grahamegrieve** [comment]
I'm not getting this. Why wouldn't you check 'contained resources'? Any contained resource? I can't see how that would be ok

**Comment by markiantorno** [comment]
This was requested by a Smile client. They want to had a flag to pass in to not check contained resources, only the containing resource.

**Comment by grahamegrieve** [comment]
That doesn't help. Why?

**Comment by markiantorno** [comment]
It's to address performance issues with their current setup. They have resources that contain hundreds and hundreds of small resources that do not require validation. This causes the validation to take a very long time when many of the 'mother' resources are passed in at a time. To improve performance, they want to have the option to to validate only the outer resource.

I'm going to just add a flag that can be passed in `validateContainedResources` which will be true by default, but they can set to false if they want to explicitly say to ignore the contained resources when validating.

**Comment by grahamegrieve** [comment]
so you want to add this to the core validator? or the cli as well>?

**Comment by markiantorno** [comment]
Just the core. CLI wouldn't include it. I want to put a flag in that can be passed up through HAPI.

**Comment by grahamegrieve** [comment]
I'll be paying particular attention to how this option is documented. Because it really sounds like a bad idea to me. "They have resources that contain hundreds and hundreds of small resources that do not require validation" - why don't they require validation?

**Comment by markiantorno** [comment]
Released as part of 5.6.2.

Let's never talk of this again.
